File: 백준/Gold/1600. 말이 되고픈 원숭이/말이 되고픈 원숭이.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer token;

    static int K,W,H;
    static int res = -1;
    static int[][] graph;
    static boolean[][][] visited;

    static int[] deltaX = {0,0,-1,1};
    static int[] deltaY = {-1,1,0,0};
    static int[] jumpX = {-2,-1,1,2,2,1,-1,-2};
    static int[] jumpY = {-1,-2,-2,-1,1,2,2,1};

    public static void main(String[] args) throws IOException {
        K = Integer.parseInt(input.readLine());
        token = new StringTokenizer(input.readLine());
        W = Integer.parseInt(token.nextToken());
        H = Integer.parseInt(token.nextToken());
        graph = new int[H][W];
        visited = new boolean[K+1][H][W];
        for (int i = 0; i < H; i++) {
            token = new StringTokenizer(input.readLine());
            for (int j = 0; j < W; j++) {
                graph[i][j] = Integer.parseInt(token.nextToken());
            }
        }
        
        if (W == 1 && H == 1) res = 0;

        Deque<int[]> deq = new ArrayDeque<>();
        deq.addLast(new int[]{0, 0, 0, 1}); // jump, W, H, cnt
        visited[0][0][0] = true;
        flag:
        while (!deq.isEmpty()) {
            int[] cur = deq.pollFirst();
            int jumpCnt = cur[0];
            int x = cur[1];
            int y = cur[2];
            int cnt = cur[3];
            for (int k = 0; k < 4; k++) {
                int dx = x + deltaX[k];
                int dy = y + deltaY[k];
                if (isValid(dx, dy)) {
                    if (dx == W-1 && dy == H-1) {
                        res = cnt;
                        break flag;
                    }
                    if (graph[dy][dx] == 0 && !visited[jumpCnt][dy][dx]) {
                        deq.addLast(new int[]{jumpCnt, dx, dy, cnt+1});
                        visited[jumpCnt][dy][dx] = true;
                    }
                }
            }
            if (jumpCnt < K){
                for (int k = 0; k < 8; k++) {
                    int dx = x + jumpX[k];
                    int dy = y + jumpY[k];
                    if (isValid(dx, dy)) {
                        if (dx == W-1 && dy == H-1) {
                            res = cnt;
                            break flag;
                        }
                        if (graph[dy][dx] == 0 && !visited[jumpCnt+1][dy][dx]) {
                            deq.addLast(new int[]{jumpCnt+1, dx, dy, cnt+1});
                            visited[jumpCnt+1][dy][dx] = true;
                        }
                    }
                }
            }
        }
        System.out.println(res);

    }
    static boolean isValid(int x, int y) {
        return x >= 0 && x < W && y >= 0 && y < H;
    }
}

File: 백준/Gold/1600. 말이 되고픈 원숭이/말이 되고픈 원숭이.py
import sys
from collections import deque

def is_valid(x,y,jump_cnt):
    return jump_cnt<=K and 0<=x<M and 0<=y<N

delta_x = [0,0,-1,1]
delta_y = [1,-1,0,0]
delta_jump_x = [-2,-2,-1,-1,1,1,2,2]
delta_jump_y = [1,-1,2,-2,2,-2,1,-1]

K = int(sys.stdin.readline())
M,N = map(int, sys.stdin.readline().split())
graph = []
for _ in range(N):
    graph.append(list(map(int, sys.stdin.readline().split())))

visited = []
for i in range(N):
    visited.append([])
    for j in range(M):
        visited[i].append([False]*(K+1))

deq = deque()
deq.append([0,0,0,0])
visited[0][0][0] = True
res = -1
while deq:
    x,y,jump_cnt,time = deq.popleft()
    if x==M-1 and y==N-1:
        res = time
        break

    for k in range(4):
        dx = x + delta_x[k]
        dy = y + delta_y[k]

        if is_valid(dx,dy,jump_cnt) and graph[dy][dx] == 0 and not visited[dy][dx][jump_cnt]:
            deq.append([dx,dy,jump_cnt,time+1])
            visited[dy][dx][jump_cnt] = True

    for k in range(8):
        dx = x + delta_jump_x[k]
        dy = y + delta_jump_y[k]
        if is_valid(dx,dy,jump_cnt+1) and graph[dy][dx] == 0 and not visited[dy][dx][jump_cnt+1]:
            deq.append([dx,dy,jump_cnt+1,time+1])
            visited[dy][dx][jump_cnt+1] = True

print(res)